### ISSUE TITLE
Passing conn to Strategies

### ISSUE THREAD

**Issue by gaiabeatrice** [pow-auth/pow_assent#117]
Hello,
Thank you very much for this great library!

Is it possible to pass `conn` parameters to the config for `Strategies`?
I need to pass some parameters to the Auth0 config so they can appear in the `authorization_params`. I was thinking about building a custom strategy for that but I can't figure out how to pass the `conn` to it.

Is it possible?

**Comment by danschultzer** [comment]
No, `conn` has [been removed](https://github.com/pow-auth/pow_assent/blob/v0.2.0/CHANGELOG.md#v020-2019-03-09) from the strategies to remove Plug dependency.

If you depend on the default PowAssent controllers, then currently the easiest would be to update the `:authorization_params` value in a plug. It can look like this:

```elixir
defmodule MyAppWeb.PowAssentAuth0Plug do
  def init(opts), do: opts

  def call(conn, _opts) do
    conn
    |> fetch_config()
    |> update_config(conn)
  end

  defp fetch_config(conn) do
    config    = Pow.Plug.fetch_config(conn)
    providers = PowAssent.Config.get_providers(config)

    {config, providers}
  end

  defp update_config({config, providers}, conn) do
    domain    = conn.params["domain"]
    params    = [scope: "openid profile email", domain: domain]
    auth0     = Keyword.put(providers[:auth0], :authorization_params, params)
    providers = Keyword.put(providers, :auth0, auth0)
    config    = Keyword.put(config, :pow_assent, providers: providers)

    Pow.Plug.put_config(conn, config)
  end
end
```

This uses the [default `:authorization_params` for Auth0](https://github.com/pow-auth/assent/blob/v0.1.4/lib/assent/strategies/auth0.ex#L23), and adds `:domain` to it as an example. You should adjust it accordingly.

This plug can then be added to the router pipeline:

```elixir
  pipeline :auth0_config do
    plug MyAppWeb.PowAssentAuth0Plug
  end

  scope "/" do
    pipe_through [:browser, :auth0_config]

    pow_routes()
    pow_assent_routes()
  end
```


The above plug assumes you define the config somwhere in `config/*.exs`, but if you feed config in through the endpoint then it'll override any custom `:pow_assent` config there.

If you use [custom controllers](https://github.com/pow-auth/pow_assent/issues/106#issuecomment-569462062) then you could just feed in the appropriate config to the [`strategy.authorization_url/1`](https://github.com/pow-auth/pow_assent/blob/v0.4.5/lib/pow_assent/plug.ex#L37) method.

I do think it should be easier to adjust the config at runtime in PowAssent though. I'll look into it.

**Comment by gaiabeatrice** [comment]
Thank you very much for the reply! I really like this solution!

**Comment by gaiabeatrice** [comment]
I noticed something interesting that may be worth reporting.

If I try to pass something in the `authorization_params` (in my case an email), the special characters get stripped and replaced by spaces.

So, for example, if I have this email `mail+mail@mail.com`, it gets replaced with `mail mail@mail.com`

**Comment by danschultzer** [comment]
That shouldn't happen as the params are encoded with [`URI.encode_query/1`](https://github.com/pow-auth/assent/blob/v0.1.5/lib/assent/strategy.ex#L117). This outputs correctly:

```elixir
iex(1)> Assent.Strategy.Auth0.authorize_url(client_id: "test", redirect_uri: "http://localhost:4000/oauth/callback", site: "http://localhost:4000", authorization_params: [scope: "openid profile email", email: "mail+mail@mail.com"])
{:ok,
 %{
   session_params: %{state: "d7ef6c85967e87fbba9311ee03ab39a2c1f52aa4b864a"},
   url: "http://localhost:4000/authorize?client_id=test&email=mail%2Bmail%40mail.com&redirect_uri=http%3A%2F%2Flocalhost%3A4000%2Foauth%2Fcallback&response_type=code&scope=openid+profile+email&state=d7ef6c85967e87fbba9311ee03ab39a2c1f52aa4b864a"
 }}
```

In the above it's correctly encoded as `mail%2Bmail%40mail.com`.

Where do you see it gets replaced with spaces? Is it in the url the user gets redirected to?

**Comment by gaiabeatrice** [comment]
I used the code snippet you posted above, in particular this is the `update_config/2` function:

```elixir
  defp update_config({config, providers}, conn) do
    auth0 = Keyword.put(providers[:auth0], :authorization_params, authorization_params(conn))

    providers = Keyword.put(providers, :auth0, auth0)
    config = Keyword.put(config, :pow_assent, providers: providers)

    Pow.Plug.put_config(conn, config)
  end

  defp authorization_params(%{params: %{"email" => email}} = _conn) when email != nil,
    do: [scope: "openid profile email", login_hint: email]
```

If I inspect the return of `authorization_params/1` the email has spaces instead of special characters.

**Comment by danschultzer** [comment]
Sounds like the email param isn't URI encoded. How do you build the path?

Depending on your setup, I recommend any of the following way to encode the value:

```elixir
Routes.pow_assent_authorization_path(conn, :new, "auth0", email: email)

path <> "?" <> URI.encode_query(email: email)

path <> "?email=" <> URI.encode(email)
```

**Comment by danschultzer** [comment]
I've merged #116 so now the above can be done like this:

```elixir
defmodule MyAppWeb.PowAssent.UpdateProviderConfig do
  def init(opts), do: opts

  def call(conn, _opts) do
    updated_config = [authorization_params: [domain: conn.params["domain"]]]

    PowAssent.Plug.merge_provider_config(conn, :auth0, updated_config)
  end
end
```
